python/whylogs/experimental/core/metrics/embedding_metric.py

```diff
@@ -0,0 +1,129 @@
+from dataclasses import dataclass
+from enum import Enum
+from io import BytesIO
+from typing import List, Optional
+
+import numpy as np
+from sklearn.metrics.pairwise import cosine_distances, euclidean_distances
+
+from whylogs.core.metrics import StandardMetric
+from whylogs.core.metrics.deserializers import deserializer
+from whylogs.core.metrics.metric_components import MetricComponent
+from whylogs.core.metrics.metrics import MetricConfig, OperationResult
+from whylogs.core.metrics.multimetric import MultiMetric
+from whylogs.core.metrics.serializers import serializer
+from whylogs.core.preprocessing import PreprocessedColumn
+from whylogs.core.proto import MetricComponentMessage
+
+# TODO: share these with NLP code
+
+
+def _serialize_ndarray(a: np.ndarray) -> bytes:
+    bio = BytesIO()
+    np.save(bio, a, allow_pickle=False)
+    return bio.getvalue()
+
+
+def _deserialize_ndarray(a: bytes) -> np.ndarray:
+    bio = BytesIO(a)
+    return np.load(bio, allow_pickle=False)
+
+
+class MatrixComponent(MetricComponent[np.ndarray]):
+    #    mtype = np.ndarray
+    type_id = 101
+
+
+@serializer(type_id=101)
+def serialize(value: np.ndarray) -> MetricComponentMessage:
+    return MetricComponentMessage(serialized_bytes=_serialize_ndarray(value))
+
+
+@deserializer(type_id=101)
+def deserialize(msg: MetricComponentMessage) -> np.ndarray:
+    return _deserialize_ndarray(msg.serialized_bytes)
+
+
+class DistanceFunction(Enum):
+    euclidean = euclidean_distances
+    cosine = cosine_distances
+
+
+@dataclass(frozen=True)
+class EmbeddingConfig(MetricConfig):
+    references: np.ndarray  # columns are reference vectors
+    labels: Optional[List[str]] = None
+    distance_fn: DistanceFunction = DistanceFunction.cosine
+
+    def __post_init__(self) -> None:
+        if len(self.references.shape) != 2:
+            raise ValueError("Embedding reference matrix must be 2 dimensional")
+
+        if self.labels:
+            if len(self.labels) == self.references.shape[1]:
+                raise ValueError(
+                    f"Number of labels ({len(self.labels)}) must match number of reference vectors ({self.references.shape[1]})"
+                )
+
+
+@dataclass
+class EmbeddingMetric(MultiMetric):
+    references: MatrixComponent
+    labels: List[str]
+    distance_fn: DistanceFunction
+
+    def __post_init__(self):
+        submetrics = {
+            f"{label}_distance": {
+                "distribution": StandardMetric.distribution.zero(),
+                "counts": StandardMetric.counts.zero(),
+                "types": StandardMetric.types.zero(),
+            }
+            for label in self.labels
+        }
+        submetrics.update(
+            {
+                "closest": {
+                    "frequent_items": StandardMetric.frequent_items.zero(),
```

Comment:
Can you add a comment to how you will use "closest" and the "labelN_distance", I think what you are doing is providing a way to monitor the distribution of the closest vectors, and then separately we have distributions of the distances for each reference?

An short example notebook would also help provide more context.

Reply:
@jamie256, happy to walk you through the design notebook I created that shows this to hear your thoughts first.

Yes, these distributions are meant to be monitored using the platform. We may want to discuss whether we want to add functions to extend that to gain a similar level of value out of whylogs-only users.